### PR TITLE
New OpenCL kernel

### PR DESCRIPTION
**Pull request by goobur**
I incorporated a lot of the changes from sgminer into CLMiner_kernel.cl. I also added an option to change THREADS_PER_HASH to 1,2,4 or 8. Depending on which card you have, and what environment you're on this can increase performance quite a bit.

This is still pretty experimental, it can require quite a bit of tuning if you change THREADS_PER_HASH to something other than 8.

### CODE REVIEW

**Comment by ddobreff** [comment]
Kernel works fine for me, the only issue is non accurate hashrate per GPU.

**Comment by goobur** [comment]
Does smurfy's patch help with that, or is it still giving you inaccurate hash rates? I mean, I see the same effect on my machine, but I don't see any reason why this kernel should cause any inaccuracies in the reported hash rate (and the computed hash rate on my pool is reporting similar rates). Perhaps due to some driver oddity, it's scheduling the kernels at the same time on all similar cards?

I'll try heavily underclocking one of my cards to see if that shows some difference between the different rates.

File: ethminer/MinerAux.h [hunk 1]
```diff
@@ -276,6 +276,19 @@ class MinerCLI
 					break;
 				}
 			}
+		else if(arg == "--cl-threads-per-hash" && i + 1 < argc) {
```
Comment (chfast) [inline ethminer/MinerAux.h]: CUDA calls it `--cuda-parallel-hash`...

Comment (goobur) [reply ethminer/MinerAux.h]: Yes, I originally thought these were different things, but it appears that you're right, and the intention for ```--cuda-parallel-hash``` is to set the number of threads per hash. I'll change it to ```--cl-parallel-hash``` and recommit. However, is this implemented for CUDA? I looked at the kernel and it seems like changing ```--cuda-parallel-hash``` doesn't affect the kernel

Comment (ddobreff) [reply ethminer/MinerAux.h]: its supposed to affect only cuda kernel, while using -G(OpenCL) cuda isn't being loaded so --cuda-parallel-hash will make no effect at all.

**Comment by cdalvaro2** [comment]
Nice job!

**Comment by mxjoe** [comment]
These changes gave a speed boost on my iMac with Radeon Pro 560. The hashrate nearly tripled to about 14 MH/s which is much closer to the 19 MH/s I get under Windows 10 (Boot Camp).

Great job! Thanks a lot!

**Comment by smurfy** [comment]
@mxjoe try playing with the ``--cl-parallel-hash``and the other parameters.

**Comment by chfast** [comment]
This is actually slower for 280X, drop from 5.6 to 4.7.

**Comment by EoD** [comment]
@goobur This introduced a regression with Mesa/Clover.

**Comment by EoD** [comment]
@goobur Regression will be fixed in #515.

**Comment by cnilsecure** [comment]
goobar can you explain how to compile claymore.bin or kernel.bin?

**Pull request merged**